src/UrlHelper.php

```diff
@@ -63,34 +78,55 @@ public function __invoke($routeName = null, array $params = [])
             $basePath = '';
         }
 
+        // Get the options to be passed to the router
+        $routerOptions = array_key_exists('router', $options) ? $options['router'] : [];
+
         if ($routeName === null) {
-            return $basePath . $this->generateUriFromResult($params, $result);
+            return $basePath . $this->generateUriFromResult($routeParams, $result, $routerOptions);
         }
 
-        if ($result) {
-            $params = $this->mergeParams($routeName, $result, $params);
+        $reuseResultParams = ! isset($options['reuse_result_params']) || (bool) $options['reuse_result_params'];
+
+        if ($result && $reuseResultParams) {
+            // Merge RouteResult with the route parameters
+            $routeParams = $this->mergeParams($routeName, $result, $routeParams);
+        }
+
+        // Generate the route
+        $path = $basePath . $this->router->generateUri($routeName, $routeParams, $routerOptions);
+
+        // Append query parameters if there are any
+        if (count($queryParams) > 0) {
+            $path .= '?' . http_build_query($queryParams);
+        }
+
+        // Append the fragment identifier
+        if ($fragmentIdentifier !== null) {
+            if (! preg_match(self::FRAGMENT_IDENTIFIER_REGEX, $fragmentIdentifier)) {
+                throw new \InvalidArgumentException('Fragment identifier must conform to RFC 3986', 400);
+            }
+
+            $path .= '#' . $fragmentIdentifier;
         }
 
-        return $basePath . $this->router->generateUri($routeName, $params);
+        return $path;
     }
 
     /**
      * Generate a URL based on a given route.
      *
      * Proxies to __invoke().
      *
-     * @param string $route
-     * @param array $params
-     * @return string
-     * @throws Exception\RuntimeException if no route provided, and no result match
-     *     present.
-     * @throws Exception\RuntimeException if no route provided, and result match is a
-     *     routing failure.
-     * @throws RouterException if router cannot generate URI for given route.
```

Comment:
We still need param/return annotations here, and possibly throws (though those could be omitted with the reference to the `__invoke()` method). I'll add those on merge.